### PR TITLE
Catch Connection Error when redis isn't up #22

### PR DESCRIPTION
**Pull request by voidpirate**
This handles the `ConnectionError` both when starting the server and if the redis server goes away during the request lifecycle.

### CODE REVIEW

File: snappass/main.py [hunk 1]
```diff
@@ -21,15 +23,27 @@
     'hour': 3600
 }
 
+def check_redis_alive(force_exit = False):
```
Comment (jparise) [inline snappass/main.py]: I believe PEP8 style says no spaces around the `=`.

File: snappass/main.py [hunk 2]
```diff
@@ -21,15 +23,27 @@
...
+    try:
+        redis_client.ping()
+    except (ConnectionError, TimeoutError) as e:
+        if force_exit:
+            print("{}".format(e))
```
Comment (jparise) [inline snappass/main.py]: Isn't this the same as `print(e)`?

**Comment by jparise** [comment]
Thanks, this is helpful! Three small bits of feedback:

1. Could you include some unit tests that cover this change?
2. Are there other places this should be checked?
3. Would it be clearer to implement this check as a function decorator?

**Comment by voidpirate** [comment]
1. Since travis handles the redis-server. What is the preferred way to handle the mock connections and simulate a failure bringing redis offline?

2. I don't believe any other routes should be checked at this time.

3. Has been converted to use a decorator. The decorator handles whether the connection is checked at server start or within a request.

File: snappass/main.py [hunk 3]
```diff
@@ -22,13 +24,29 @@
 }
 
 
+def check_redis_alive(fn):
+    def inner(*args, **kwargs):
+        try:
+            redis_client.ping()
+        except ConnectionError as e:
+            print(e)
```
Comment (nichochar) [inline snappass/main.py]: Nice, how about also printing a friendly error message. I'm debating if we need _only_ a pretty error message like `"Failed to connect to redis"`.
I don't mind using message + stacktrace, but let's make it user friendly.

File: snappass/main.py [hunk 4]
```diff
@@ -22,13 +24,29 @@
...
+            if fn.__name__ == "main":
+                sys.exit(0)
+            else:
+                return abort(500)
```
Comment (nichochar) [inline snappass/main.py]: This is good for now, but down the road we may want to try and add a [custom error template](http://flask.pocoo.org/docs/0.11/patterns/errorpages/) with a message explaining the error.

File: snappass/main.py [hunk 5]
```diff
@@ -22,13 +24,29 @@
 }
 
 
+def check_redis_alive(fn):
+    def inner(*args, **kwargs):
+        try:
+            redis_client.ping()
```
Comment (nichochar) [inline snappass/main.py]: Note that with this `ping()` design, it is still possible (even if, admittedly, unlikely), that the ping works but the function call below still fails to a `ConnectionError`.

Why not simply wrap the function call`fn(*args, **kwargs)` in the `try/except` ?

**Comment by nichochar** [comment]
Awesome! Thanks for making the change @voidpirate 

I also had some minor suggestions in the comments.

### Please address before merge
* try calling the function in the `try/except` rather than the `ping()`

### Won't block merge
* would be nice to have a custom error template
* tests

@jparise the reason I put the testing in the optional here is that I think we should actually mock the redis server entirely. I created an [issue](https://github.com/pinterest/snappass/issues/25)


It would be nice to address the "calling the function" instead of the ping.

File: snappass/main.py [hunk 6]
```diff
@@ -22,13 +24,29 @@
...
+        except ConnectionError as e:
+            print(e)
+            if fn.__name__ == "main":
```
Comment (nichochar) [inline snappass/main.py]: Minor nit: can we switch this to single quotes? I haven't done it yet, but I am going to write a coding style for this repo, and strings favor single quotes if possible.

**Comment by voidpirate** [comment]
I've made the changes. Thank you very much for the feedback :)

The fails on travis is actually not due to any of the code changes, but failed to install `Python 3.5` (see attached).
screen shot 2016-08-22 at 10 05 59 pm

**Comment by nichochar** [comment]
Kicked the travis job back, and it's now successful. 
Thanks for making all the changes @voidpirate !

**Pull request merged**